### PR TITLE
Move source files to `src/` and objects to `obj/`

### LINKED ISSUE #44: Should we move source files to `src/`?
The root directory is quite crowded. Should we move source and header files (except maybe the generated `config.h`) to `src/`, like the `exec/` directory for shell scripts?

If so, we should do it before #9 is merged, or maybe in #9.

### PR DESCRIPTION
**Pull request by GRFreire**
The root directory is quite crowded. This pr makes so all source file are inside the `src/` dir and all objects are inside `obj/`.

The `Makefile` still is in the root directory, as well as the `config.h` for easy configuration.

Closes #44

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -33,8 +33,10 @@ CPPFLAGS = -D_XOPEN_SOURCE=700 \
 
 LDLIBS = -lImlib2 -lX11 -lXft -lfontconfig $(OPTIONAL_LIBS)
 
-OBJS = autoreload_$(AUTORELOAD).o commands.o image.o main.o options.o \
-  thumbs.o util.o window.o
+SRCS = src/autoreload_$(AUTORELOAD).c src/commands.c src/image.c src/main.c src/options.c \
+  src/thumbs.c src/util.c src/window.c
+
+OBJS = $(patsubst src/%.c, obj/%.o, $(SRCS))
```
Comment (N-R-K) [inline Makefile:39]: I'm not sure if this is POSIX, and we're trying to keep things POSIX compliant. Will need to look into this.

File: Makefile [hunk 2]
```diff
@@ -46,11 +48,11 @@ nsxiv: $(OBJS)
 	@echo "LINK $@"
 	$(CC) $(LDFLAGS) -o $@ $^ $(LDLIBS)
 
-.c.o:
+obj/%.o: src/%.c
```
Comment (N-R-K) [inline Makefile:51]: Same thing, need to look into if this is POSIX or not.

**Review by N-R-K — CHANGES_REQUESTED** [review]
Do we want to keep config.h and config.def.h in root? Imo we should move it to src, cleaner that way.

File: Makefile (hunk 2 above)
Comment (N-R-K) [reply Makefile:51]: > Implementations may allow other characters in target names as extensions. The interpretation of targets containing the characters '%' and '"' is implementation-defined.

Okay, so this is certainly not posix.

**Comment by GRFreire** [comment]
> Do we want to keep config.h and config.def.h in root? Imo we should move it to src, cleaner that way.

I think we want. It is the default location of most programs that have a `config.h` (i think). Although we are moving the source files to a separate directory, keeping the `config.h` in the root dir makes it easy to "new" people (both familiar with `suckless` and mainline `sxiv`) to realize that there is a `config.h` to hack on.

**Comment by GRFreire** [comment]
Is there a specification on POSIX compliant make? I couldn't find one. GNU Make does not have any flags you can pass to disable all gnu extensions, but adding `.POSIX:` at the beginning should do the job. If it works, so the current `makefile` is posix and no need to change (regarding this matter).

**Comment by N-R-K** [comment]
> adding `.POSIX:` at the beginning should do the job

Afaik, that doesn't do what you think it's doing. `.POSIX` only makes sure that extensions which conflicts with posix are handled in a posix way, it doesn't disable non-posix extensions.

**Comment by GRFreire** [comment]
I am having some trouble solving the conflicts (or maybe trouble with github). In my machine I can easily resolve conflicts and fix, but when I push to github, it still says that there are conflicts and I cannot look at the web "editor".

Could some nice individual try to help me with that? I've tried twice (the only two ways I know) and didn't work.

**Comment by GRFreire** [comment]
> Afaik, that doesn't do what you think it's doing. .POSIX only makes sure that extensions which conflicts with posix are handled in a posix way, it doesn't disable non-posix extensions.

Ok, I think we should try and get other `make` program with no extension then.

**Comment by N-R-K** [comment]
> Ok, I think we should try and get other `make` program with no extension then.

I'm not sure if such make exists. You can read the posix man page though, `man 1p make`.

**Comment by N-R-K** [comment]
I don't think there's any POSIX alternative to `%`
Which means we'll have to manually write the rule for each src/obj.

**Comment by GRFreire** [comment]
Since apparently there is no `make` with no extensions, maybe all majors `make` variants (gnu, the one used on openBSD, netBSD, etc) support `%` (as well as other gnu extensions that we are using). I'll take a look into that tomorrow.

**Comment by N-R-K** [comment]
Currently, there's 3 commits on `sxiv` trying to make the `Makefile` POSIX compliant, so clearly that has some value. If we can't sanely do this without using non-POSIX extensions, then imo we might just be better off dropping it.

Since this is *mostly* a cosmetic thing, I don't think it's worth breaking POSIX just for it.

**Comment by GRFreire** [comment]
Ok, I agree that we should not break POSIX. I can see 3 options:
- Just close this pr. Simple, but keeps all problems stated in #44 
- Keep all source files in `src/`, as well as POSIX. It fixes all "cosmetic problems" but the functionality ones still remain.
- Hard code both `SRCS` and `OBJS` instead of using `patsubst` and/or `%`. Example illustrated below. Fixes all the problems talked in the issue, the problem is that every time one wants to change `SRCS` it would be necessary to manually change `OBJS`. Shouldn't be such of a problem, given that this don't change as much.

Example:
```diff
SRCS = src/autoreload_$(AUTORELOAD).c src/commands.c src/image.c src/main.c src/options.c \
  src/thumbs.c src/util.c src/window.c

- OBJS = $(patsubst src/%.c, obj/%.o, $(SRCS))

+ OBJS = obj/autoreload_$(AUTORELOAD).o obj/commands.o obj/image.o obj/main.o obj/options.o \
+   obj/thumbs.o obj/util.o obj/window.o
```

I personally prefer the last approach, but I want to hear what you guys have to say.

**Comment by N-R-K** [comment]
@GRFreire I'm okay with hardcoding `OBJS` since we dont have too many sources. But how do you deal with `%` issue?

**Comment by GRFreire** [comment]
I was able to change the `Makefile` so it meets 3rd idea (https://github.com/nsxiv/nsxiv/pull/81#issuecomment-925120801).

> @GRFreire I'm okay with hard-coding OBJS since we dont have too many sources. But how do you deal with % issue?

Dealt with in this last commit. Don't know if is the best solution, but it is working.

**Comment by GRFreire** [comment]
> Not quite. Build fails on my end on "version.h"

Mine was there from before the update. Looks like it is an easy fix.

> This also breaks make's dependency tracking... In other words, it'll trigger a full build everytime make is run.

Ummm, thats a tricky one. I think that this is happening because the items inside `$(SRCS)` do not "point" to a file.

**Comment by GRFreire** [comment]
Now it should compile with no problems. Please tell me otherwise.

Also, the `Makefile` is a bit cleaner now and the redundancy of source files in `SRCS` and `OBJS` was removed.

**Comment by N-R-K** [comment]
Dependency tracking is still broken. It compiles the first time, but if you edit a single file, lets say options.c and then run `make` it should rebuild `option.c` and relink. But it doesn't do anything.

**Comment by N-R-K** [comment]
The only way I've found to do `src/` and `obj/` in a POSIX way is to hardcode rules for every single `src/*.c` files. Imo it's not worth creating that mess just for cosmetic purposes.

**Comment by GRFreire** [comment]
I suspect you're right. I've come up to 5/6 ideas on how to implement this but all breaks posix or `make` dont track files nicely.

Closing this, if anyone else have a solution feel free to open it up again.

**Pull request closed without merging**